tests/math_tools/test_curl.cpp

```diff
@@ -0,0 +1,134 @@
+            GridRho::init<GridRho>(Coord<Rho>(0.0), Coord<Rho>(1.0), nrho));
+    ddc::init_discrete_space<GridTheta>(GridTheta::init<GridTheta>(
+            build_uniform_break_points(Coord<Theta>(0.0), Coord<Theta>(2 * M_PI), ntheta)));
+    ddc::init_discrete_space<GridPhi>(
+            GridPhi::init<GridPhi>(Coord<Phi>(0.0), Coord<Phi>(2 * M_PI), nphi));
+
+    Coord<Rho, Theta, Phi> coord
+            = ddc::coordinate(Idx<GridRho, GridTheta, GridPhi>(nrho - 1, ntheta - 5, 0));
+    DTensor<CovBasisSpatial, CovBasisSpatial> grad_B;
+
+    double const r = ddc::get<Rho>(coord);
+    double const theta = ddc::get<Theta>(coord);
+    // Compute dB_r/dr, dB_theta/dr, dB_r/dtheta, dB_theta/dtheta
+    ddcHelper::get<Rho_cov, Rho_cov>(grad_B) = R0 * std::cos(theta) * std::sinh(r);
+    ddcHelper::get<Rho_cov, Theta_cov>(grad_B) = -R0 * std::sin(theta) * std::cosh(r);
+    ddcHelper::get<Rho_cov, Phi_cov>(grad_B) = 0.0;
+    ddcHelper::get<Theta_cov, Rho_cov>(grad_B) = -R0 * std::sin(theta) * std::cosh(r);
+    ddcHelper::get<Theta_cov, Theta_cov>(grad_B) = -R0 * std::cos(theta) * std::sinh(r);
+    ddcHelper::get<Theta_cov, Phi_cov>(grad_B) = 0.0;
+    ddcHelper::get<Phi_cov, Rho_cov>(grad_B) = 0.0;
+    ddcHelper::get<Phi_cov, Theta_cov>(grad_B) = 0.0;
+    ddcHelper::get<Phi_cov, Phi_cov>(grad_B) = 0.0;
+
+    DVector<Rho, Theta, Phi> curl_B = curl(grad_B, coord);
+
+    double eps = 1e-12;
+    EXPECT_NEAR(ddcHelper::get<Rho>(curl_B), 0.0, eps);
+    EXPECT_NEAR(ddcHelper::get<Theta>(curl_B), 0.0, eps);
+    EXPECT_NEAR(ddcHelper::get<Phi>(curl_B), 0.0, eps);
+}
+
+// Test the curl for a simple B-field where the
+// result can be analytically computed.
+// Let's use B = (0, 0, 1/(R0+r))
+TEST(CurlTests, specific_field_and_point)
+{
+    using Mapping2D = CircularToCartesian<Rho, Theta, R, Z>;
+    using ToroidalMapping = ToroidalToCylindrical<Mapping2D, Zeta, Phi>;
+    using CylindricalMapping = CylindricalToCartesian<R, Z, Zeta, X, Y>;
+
+    double R0 = 6.2;
+    Coord<R, Z> origin_point(R0, 0.0);
+    Mapping2D polar_to_RZ(origin_point);
+    ToroidalMapping toroidal_to_cylindrical(polar_to_RZ);
+    CylindricalMapping cylindrical_to_cartesian;
+    CombinedMapping<CylindricalMapping, ToroidalMapping, Coord<Rho, Theta, Phi>>
+            mapping(cylindrical_to_cartesian, toroidal_to_cylindrical);
+    Curl curl(mapping);
+
+    using BasisSpatial = VectorIndexSet<Rho, Theta, Phi>;
+    using CovBasisSpatial = get_covariant_dims_t<BasisSpatial>;
+    IdxStepRho nrho(10);
+    IdxStepTheta ntheta(10);
+    IdxStepPhi nphi(2);
+    ddc::init_discrete_space<GridRho>(
+            GridRho::init<GridRho>(Coord<Rho>(0.0), Coord<Rho>(1.0), nrho));
+    ddc::init_discrete_space<GridTheta>(GridTheta::init<GridTheta>(
+            build_uniform_break_points(Coord<Theta>(0.0), Coord<Theta>(2 * M_PI), ntheta)));
+    ddc::init_discrete_space<GridPhi>(
+            GridPhi::init<GridPhi>(Coord<Phi>(0.0), Coord<Phi>(2 * M_PI), nphi));
```

Comment:
Are these grids used?

Reply:
I actually do not understand how it works. I only need a valid coordinate (point).